Lib/urllib/parse.py

```diff
@@ -443,6 +451,7 @@ def urlsplit(url, scheme='', allow_fragments=True):
             if '?' in url:
                 url, query = url.split('?', 1)
             _checknetloc(netloc)
+            url = _remove_unsafe_bytes_from_url(url)
```

Comment:
I think we need to do this up front right after `_coerce_args(url, ...)` (and should do the same in 3.9 and 3.10).

by this point in 3.8 we've potentially allowed characters to slip through into query and fragment of http: urls.

at a minumum if we weren't do this right after _coerce_args before looking in the cache, this needs to be done right before _any_ splitting happens in this branch of code.

Reply:
the test case should be updated to include an invalid character in each of the five portions of the url.

Reply:
> allowed characters to slip through into query and fragment

I had a doubt initially, and I thought (from practice) that newlines and tabs are percent-encoded when in query/fragment than they are removed upfront.

We will have find a spec that will unambiguously state what should be done.

Reply:
We've already got that unambiguous spec.  https://url.spec.whatwg.org/#concept-basic-url-parser step 3.  strip these three characters no matter what before the parsing state machine starts.